DOCSP-37538 Specify a Query (#51)
mongodb/docs-pymongo: source/specify-a-query.txt

.. _pymongo-specify-query:

===============
Specify a Query
===============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: expressions, operations, read, write, filter

Overview
--------

> In this guide, you can learn how to specify a query by using {+driver-short+}.

Author: Good point, this is especially true when PyMongo queries are largely the same as mongosh queries (whereas drivers like C# or Java specify them differently). I added a link at the bottom of the page to the manual docs for learning more

You can refine the set of documents that a query returns by creating a
**query filter**. A query filter is an expression that specifies the search
criteria MongoDB uses to match documents in a read or write operation.
In a query filter, you can prompt the driver to search for documents with an
exact match to your query, or you can compose query filters to express more
complex matching criteria.

Sample Data
~~~~~~~~~~~

The examples in this guide run operations on a collection called
``fruits`` that contains the following documents:

.. code-block:: json

   { "_id": 1, "name": "apples", "qty": 5, "rating": 3, "color": "red", "type": ["fuji", "honeycrisp"] },
   { "_id": 2, "name": "bananas", "qty": 7, "rating": 4, "color": "yellow", "type": ["cavendish"] },
   { "_id": 3, "name": "oranges", "qty": 6, "rating": 2, "type": ["naval", "mandarin"] },
   { "_id": 4, "name": "pineapple", "qty": 3, "rating": 5, "color": "yellow" },

The following code example shows how to create a database and collection, then
insert the sample documents into your collection:

> .. literalinclude:: /includes/specify-query/specify-queries.py

Author: This is here to make it easy for beginners that want test the queries themselves. Since this page isn't using the Atlas sample data, having this code block makes it easier for them to load the data that the examples are using.

   :start-after: start-sample-data
   :end-before: end-sample-data
   :language: python
   :copyable:

Exact Match
-----------

Literal value queries return documents with an exact match to your query filter.

The following example specifies a query filter as a parameter to the ``find()``
method. The code returns all documents with a ``color`` field value of ``"yellow"``:

.. io-code-block::
   :copyable: 

   .. input:: /includes/specify-query/specify-queries.py
      :start-after: start-find-exact
      :end-before: end-find-exact
      :language: python

   .. output::
      
      {'_id': 2, 'name': 'bananas', 'qty': 7, 'rating': 4, 'color': 'yellow', 'type': ['cavendish']}
      {'_id': 4, 'name': 'pineapple', 'qty': 3, 'rating': 5, 'color': 'yellow'}

.. tip:: Find All Documents

>    To find all documents in a collection, call the ``find()`` method without

Author: Thanks for pointing that out. Changed to pass an empty dict

      :end-before: end-find-element
      :language: python

   .. output::

      {'_id': 1, 'name': 'apples', 'qty': 5, 'rating': 3, 'color': 'red', 'type': ['fuji', 'honeycrisp']}
      {'_id': 2, 'name': 'bananas', 'qty': 7, 'rating': 4, 'color': 'yellow', 'type': ['cavendish']}
      {'_id': 4, 'name': 'pineapple', 'qty': 3, 'rating': 5, 'color': 'yellow'}

Evaluation Operators
--------------------

Evaluation operators return data based on evaluations of either individual
fields or the entire collection's documents.

The following is a list of common evaluation operators:

- ``$text``, which performs a text search on the documents
- ``$regex``, which returns documents that match a specified regular expression
- ``$mod``, which performs a modulo operation on the value of a field and
  returns documents where the remainder is a specified value

To view a full list of evaluation operators, see the :manual:`Evaluation Query Operators
</reference/operator/query-evaluation/>` guide in the MongoDB Server manual.

The following example specifies an evaluation operator in a query filter as a
parameter to the ``find()`` method. The code returns all documents with a

> ``name`` field that matches the regular expression ``/apple/``:

Reviewer: How come you're using `r"apple"` instead of `/apple/`? The latter looks more like a regex. 

I like the following example: `list(clxn.find({ "name" : { "$regex" : "p{2,}" }} ))` which will give you any match that has at least 2 consequence p characters.

Author: `r"apple` was just the way it worked when I was testing the queries, but I do like yours better as it's probably a more realistic usage of regex, but still simple enough to understand. Updated

.. literalinclude:: /includes/specify-query/specify-queries.py
   :start-after: start-sample-data
   :end-before: end-sample-data
   :language: python
   :copyable:

Exact Match
-----------

Literal value queries return documents with an exact match to your query filter.

The following example specifies a query filter as a parameter to the ``find()``
method. The code returns all documents with a ``color`` field value of ``"yellow"``:

.. io-code-block::
   :copyable: 

   .. input::
      :language: python

      results = collection.find({ "color": "yellow" })

   .. output::
      
      {'_id': 2, 'name': 'bananas', 'qty': 7, 'rating': 4, 'color': 'yellow', 'type': ['cavendish']}
      {'_id': 4, 'name': 'pineapple', 'qty': 3, 'rating': 5, 'color': 'yellow'}

.. tip:: Find All Documents

   To find all documents in a collection, call the ``find()`` method without
   any parameters. The following example finds all documents in a
   collection:

   .. literalinclude:: /includes/specify-query/specify-queries.py
      :start-after: start-find-all
      :end-before: end-find-all
      :language: python
      :copyable:

Comparison Operators
--------------------

> Comparison operators analyze the value in a document against the specified value
> in your query filter. The following is a list of common comparison operators:

Reviewer 2: ```suggestion
Comparison operators evaluate a document field value against a specified value
in your query filter. The following is a list of common comparison operators:
```

</reference/operator/query-array/>` guide in the MongoDB Server manual.

The following example specifies an array operator in a query filter as a
parameter to the ``find()`` method. The code returns all documents with a
``type`` array field containing ``2`` elements:

.. io-code-block::
   :copyable:

   .. input::
      :language: python

      results = collection.find({
          "type" : { "$size": 2 }
      })

      for f in results:
          print(f)

   .. output::

      {'_id': 1, 'name': 'apples', 'qty': 5, 'rating': 3, 'color': 'red', 'type': ['fuji', 'honeycrisp']}
      {'_id': 3, 'name': 'oranges', 'qty': 6, 'rating': 2, 'type': ['naval', 'mandarin']}

Element Operators
-----------------

Element operators query data based on the presence or type of a field.

To learn more about element operators, see the :manual:`Element Query Operators
</reference/operator/query-element/>` guide in the MongoDB Server manual.

The following example specifies an element operator in a query filter as a
parameter to the ``find()`` method. The code returns all documents that have a
``color`` field:

.. io-code-block::
   :copyable:

   .. input::
      :language: python

>       results = collection.find( {"color" : { "$exists": "true" }} )

Reviewer 2: S: run formatter on all code examples if possible
```suggestion
      results = collection.find( { "color" : { "$exists": "true" }} )
```

Overview
--------

In this guide, you can learn how to specify a query by using {+driver-short+}.

You can refine the set of documents that a query returns by creating a
**query filter**. A query filter is an expression that specifies the search
criteria MongoDB uses to match documents in a read or write operation.
In a query filter, you can prompt the driver to search for documents with an
exact match to your query, or you can compose query filters to express more
complex matching criteria.

Sample Data
~~~~~~~~~~~

The examples in this guide use the following documents in a collection called
``fruits``:

.. code-block:: json

   { "_id": 1, "name": "apples", "qty": 5, "rating": 3, "color": "red", "type": ["fuji", "honeycrisp"] },
   { "_id": 2, "name": "bananas", "qty": 7, "rating": 4, "color": "yellow", "type": ["cavendish"] },
   { "_id": 3, "name": "oranges", "qty": 6, "rating": 2, "type": ["naval", "mandarin"] },
   { "_id": 4, "name": "pineapple", "qty": 3, "rating": 5, "color": "yellow" },

The following code example shows how to create a database and collection, and
insert the sample documents into your collection:

> .. code-block:: python

Reviewer 2: Applies to all sections!

> .. code-block:: python

Author: I moved the code blocks that aren't `io-code-blocks` into a separate file

> .. code-block:: python

Reviewer 2: You can! See [this](https://github.com/mongodb/docs-rust/blob/dee0bd55d7f188f41e77c890a46be734aedcb643/source/fundamentals/crud/read-operations/query.txt#L173-L188) for reference

> .. code-block:: python

Author: Oh wow! I didn't know they finally added it. Updated

   from pymongo import MongoClient

>    uri = "<Your Connection URI>"

Author: Good call, I took that wording from another repo. Changed to match what we have in the Get Started section of this repo

   client = MongoClient(uri)

   try:
       database = client["sample_fruit"]
       collection = database["fruits"]

       collection.insert_many([
           { "_id": 1, "name": "apples", "qty": 5, "rating": 3, "color": "red", "type": ["fuji", "honeycrisp"] },
           { "_id": 2, "name": "bananas", "qty": 7, "rating": 4, "color": "yellow", "type": ["cavendish"] },
           { "_id": 3, "name": "oranges", "qty": 6, "rating": 2, "type": ["naval", "mandarin"] },
           { "_id": 4, "name": "pineapple", "qty": 3, "rating": 5, "color": "yellow" },
       ])

       client.close()

   except Exception as e:
      raise Exception("Unable to insert the documents due to the following error: ", e)

Exact Match
-----------

Literal value queries return documents with an exact match to your query filter.

The following example specifies a query filter as a parameter to the ``find()``
method. The method finds all documents with a ``color`` field value of ``"yellow"``:

.. code-block::
   :copyable:

>    results = collection.find({ "color": "yellow" })

Reviewer 2: S: include output for all code examples as done [here](https://www.mongodb.com/docs/drivers/rust/current/fundamentals/crud/read-operations/query/)

Author: Included in all but the "find all" note (since it shows all the documents right before that)

.. tip:: Find All Documents

   To find all documents in a collection, call the ``find()`` method without
   passing any arguments. The following example finds all documents in a collection:

   .. code-block:: python

      results = collection.find()

Comparison Operators
--------------------

Comparison operators analyze the value in a document against the specified value
in your query filter. The following is a list of common comparison operators:

- Greater than: ``$gt``
- Less than or Equal: ``$lte``

> - Not equal: ``$ne``

Reviewer 2: S: reverse order of the operator vs description because of the way the list is introduced

> For a full list of comparison operators, see the :manual:`Comparison Query Operators

Reviewer 2: ```suggestion
To view a full list of comparison operators, see the :manual:`Comparison Query Operators
```

</reference/operator/query-comparison/>` guide in the MongoDB Server manual.

> The following example specifies a comparison operator in a query filter as a
> parameter to the ``find()`` method. The method finds all documents with a
> ``rating`` field value greater than ``2``:

Reviewer 2: ```suggestion
The following example specifies a comparison operator in a query filter as a
parameter to the ``find()`` method. The code returns all documents with a
``rating`` field value greater than ``2``:
```

.. io-code-block::
   :copyable: 

   .. input::
      :language: python

      results = collection.find({"rating": {"$gt" : 2}})

      for f in results:
          print(f) 

   .. output::

      {'_id': 1, 'name': 'apples', 'qty': 5, 'rating': 3, 'color': 'red', 'type': ['fuji', 'honeycrisp']}
      {'_id': 2, 'name': 'bananas', 'qty': 7, 'rating': 4, 'color': 'yellow', 'type': ['cavendish']}
      {'_id': 4, 'name': 'pineapple', 'qty': 3, 'rating': 5, 'color': 'yellow'}

Logical Operators
-----------------

Logical operators match documents using logic applied to the results of two or
more sets of expressions. The following is a list of logical operators: 

- ``$and``, which returns all documents that match the conditions of *both* clauses
- ``$or``, which returns all documents that match the conditions of *either* clause
- ``$nor``, which returns all documents that *do not* match the conditions of either clause

> - ``$not``, which returns all documents that *do not* match the expression

Author: Good catch

For more information about logical operators, see the :manual:`Logical Query Operators

> </reference/operator/query-logical/>` guide in the MongoDB Server manual.

Reviewer 2: Doesn't seem resolved in all instances

> </reference/operator/query-logical/>` guide in the MongoDB Server manual.

Author: Find and replace failed me

The following example specifies a logical operator in a query filter as a
parameter to the ``find()`` method. The method finds all documents with a
``qty`` field value greater than ``5`` **or** a ``color`` field value of
``"yellow"``:

.. io-code-block::
   :copyable:

   .. input::
      :language: python

      results = collection.find({ 
          "$or": [
                { "qty": { "$gt": 5 }},
                { "color": "yellow" }
          ]
      })

      for f in results:
          print(f)

   .. output::

      {'_id': 2, 'name': 'bananas', 'qty': 7, 'rating': 4, 'color': 'yellow', 'type': ['cavendish']}
      {'_id': 3, 'name': 'oranges', 'qty': 6, 'rating': 2, 'type': ['naval', 'mandarin']}
      {'_id': 4, 'name': 'pineapple', 'qty': 3, 'rating': 5, 'color': 'yellow'}

Array Operators
---------------

Array operators match documents based on the value or quantity of elements in an

> array field. The following is a list of array operators:

Reviewer 2: Q: is this a complete list?

Author: Yes this is all of the array operators. I slightly reworded to make that clear

To learn more about element operators, see the :manual:`Element Query Operators
</reference/operator/query-element/>` guide in the MongoDB Server manual.

The following example specifies an element operator in a query filter as a
parameter to the ``find()`` method. The method finds all documents that have a
``color`` field:

.. io-code-block::
   :copyable:

   .. input::
      :language: python

      results = collection.find( {"color" : { "$exists": "true" }} )

      for f in results:
          print(f)

   .. output::

      {'_id': 1, 'name': 'apples', 'qty': 5, 'rating': 3, 'color': 'red', 'type': ['fuji', 'honeycrisp']}
      {'_id': 2, 'name': 'bananas', 'qty': 7, 'rating': 4, 'color': 'yellow', 'type': ['cavendish']}
      {'_id': 4, 'name': 'pineapple', 'qty': 3, 'rating': 5, 'color': 'yellow'}

Evaluation Operators
--------------------

Evaluation operators return data based on evaluations of either individual
fields or the entire collection's documents.

The following is a list of common evaluation operators:

- ``$text``, which performs a text search on the documents
- ``$regex``, which returns documents that match a specified regular expression

> - ``$mod``, which performs a modulo operation on the value of a field and
>   returns documents with a specified result

Reviewer 2: ```suggestion
- ``$mod``, which performs a modulo operation on the value of a field and
  returns documents where the remainder is a specified value
```
